artemis-protocols/artemis-amqp-protocol/src/main/java/org/apache/activemq/artemis/protocol/amqp/broker/AMQPMessage.java

```diff
@@ -752,31 +771,31 @@ protected ReadableBuffer createCopyWithNewDeliveryCount(int deliveryCount) {
       // otherwise we want to write the original header if present.  When a
       // Header is present we need to copy it as we are updating the re-delivered
       // message and not the stored version which we don't want to invalidate here.
-      Header header = this.header;
-      if (header == null) {
-         header = new Header();
+      Header localHeader = this.header;
+      if (localHeader == null) {
+         localHeader = new Header();
       } else {
-         header = new Header(header);
+         localHeader = new Header(header);
       }
 
-      header.setDeliveryCount(UnsignedInteger.valueOf(amqpDeliveryCount));
+      localHeader.setDeliveryCount(UnsignedInteger.valueOf(amqpDeliveryCount));
       TLSEncode.getEncoder().setByteBuffer(new NettyWritable(result));
-      TLSEncode.getEncoder().writeObject(header);
+      TLSEncode.getEncoder().writeObject(localHeader);
       TLSEncode.getEncoder().setByteBuffer((WritableBuffer) null);
 
-      writeDeliveryAnnotationsForSendBuffer(result);
+      writeDeliveryAnnotationsForSendBuffer(result, deliveryAnnotations);
       // skip existing delivery annotations of the original message
-      getData().position(encodedHeaderSize + encodedDeliveryAnnotationsSize);
-      result.writeBytes(getData().byteBuffer());
-      getData().position(0);
+      duplicate.position(encodedHeaderSize + encodedDeliveryAnnotationsSize);
+      result.writeBytes(duplicate.byteBuffer());
 
       return new NettyReadable(result);
    }
 
-   protected void writeDeliveryAnnotationsForSendBuffer(ByteBuf result) {
-      if (deliveryAnnotationsForSendBuffer != null && !deliveryAnnotationsForSendBuffer.getValue().isEmpty()) {
+   protected void writeDeliveryAnnotationsForSendBuffer(ByteBuf result, DeliveryAnnotations deliveryAnnotations) {
+      DeliveryAnnotations daToUse = deliveryAnnotations != null ? deliveryAnnotations : deliveryAnnotationsForSendBuffer;
```

Comment:
This method is called a few levels deep during encode/send, and can then seem like it wont use the arg that was passed all the way through as the value to write. For clarity I think this should just use the passed in value, with the decision of what to send already being made at the start of the process in the getSendBuffer method (already the case, suggesting this bit currently doesnt do anything except confuse things).

Reply:
done.. will be on my next commit